File: app/src/main/java/com/example/estres2/ui/registros/FragmentRegistro.java
package com.example.estres2.ui.registros;

import android.content.Context;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.core.content.res.ResourcesCompat;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.RecyclerView.Adapter;

import android.os.Environment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.example.estres2.MenuPrincipal;
import com.example.estres2.R;
import com.example.estres2.UsuarioBoleta;
import com.example.estres2.almacenamiento.database.DB;
import com.example.estres2.almacenamiento.entidades.archivo.Archivo;
import com.example.estres2.almacenamiento.entidades.usuario.Usuario;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class FragmentRegistro extends Fragment {
    private final ArrayList<ListaRegistro> lRegistro = new ArrayList<>();
    public Context mContext;
    public Usuario user;

    public FragmentRegistro() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public void onAttach(@NonNull Context context) {
        // TODO Auto-generated method stub
        super.onAttach(context);
        mContext = context;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        List<String> arregloArchivos = obtenerRegistros();
        if (!arregloArchivos.isEmpty()){
            for (int i = 0; i <= arregloArchivos.size() - 1; i++ ) {
                lRegistro.add(new ListaRegistro(arregloArchivos.get(i), R.drawable.ic_registros_menu));
            }
        } else {
            Toast.makeText(mContext, "Carpeta Vacia", Toast.LENGTH_LONG).show();
            lRegistro.add(new ListaRegistro("Carpeta Vacia", R.drawable.ic_sin_registro));
        }

        View root = inflater.inflate(R.layout.fragment_registro, container, false);
        RecyclerView mRecyclerView = root.findViewById(R.id.Resgistros);
        mRecyclerView.setHasFixedSize(true);
        RecyclerView.LayoutManager mLayaoutManager = new LinearLayoutManager(this.getContext());
        Adapter<ListaRegistroAdapter.ListaRegistroViewHolder> mAdapter = new ListaRegistroAdapter(lRegistro, getContext(), user.getBoleta());
        mRecyclerView.setLayoutManager(mLayaoutManager);
        mRecyclerView.setAdapter(mAdapter);

        DividerItemDecoration itemDecoration = new DividerItemDecoration(mContext, DividerItemDecoration.VERTICAL);
        itemDecoration.setDrawable(Objects.requireNonNull(ResourcesCompat.getDrawable(getResources(), R.drawable.divider, null)));
        mRecyclerView.addItemDecoration(itemDecoration);
        return root;
    }

    public List<String> obtenerRegistros() {
        List<String> item = new ArrayList<>();
        DB db = new DB(getContext());
        if (!(getActivity() == null)) {
            user = UsuarioBoleta.INSTANCE.getObjectBoleta();
            File Carpeta = new File(Environment.getExternalStorageDirectory() + "/Monitoreo" + user.getBoleta());
            if (Carpeta.exists()) {
                File[] files = Carpeta.listFiles();
                assert files != null;
                for (File file : files) {
                    //Sacamos del array files un fichero
                    if(file.getPath().endsWith(".csv") && db.ConsultarArchivo(new Archivo(file.getName(), user.getBoleta())))
                        item.add(file.getName());
                }
                return item;
            } else {
                Toast.makeText(mContext, "No existe la carpeta del usuario", Toast.LENGTH_LONG).show();
            }
        } else {
            Log.d("getActivity", "La actividad es nulla");
        }
        item.add("Carpeta vacia");
        return item;
    }
}

File: app/src/main/java/com/example/estres2/ui/registros/FragmentRegistro.kt
package com.example.estres2.ui.registros

import android.content.Context
import android.os.Bundle
import android.os.Environment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.GridLayoutManager
import com.example.estres2.R
import com.example.estres2.UsuarioBoleta.getObjectBoleta
import com.example.estres2.almacenamiento.database.DB
import com.example.estres2.almacenamiento.entidades.archivo.Archivo
import com.example.estres2.almacenamiento.entidades.usuario.Usuario
import com.example.estres2.databinding.FragmentRegistroBinding
import com.example.estres2.ui.viewmodel.MenuViewModel
import java.io.File

class FragmentRegistro: Fragment() {
    private var _binding: FragmentRegistroBinding? = null
    private val binding get() = _binding!!
    private val menuViewModel: MenuViewModel by viewModels()
    private lateinit var mContext: Context
    private lateinit var user: Usuario
    private val lRegistro: MutableList<ListaRegistro> = ArrayList()

    override fun onCreateView(
            inflater: LayoutInflater, container: ViewGroup?,
            savedInstanceState: Bundle?,
    ): View {
        _binding = FragmentRegistroBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mContext = binding.root.context
        setObservers()
        showRegister()
    }

    override fun onDestroy() {
        super.onDestroy()
        _binding = null
    }

    private fun setObservers() {
        menuViewModel.apply {
            updateRegisters.observe(viewLifecycleOwner) {
                when (it) {
                    true -> {
                        Toast.makeText(context, "Se actualizo el estado del observador: $it", Toast.LENGTH_LONG).show()
                        showRegister()
                    }
                }
            }
        }
    }

    private fun showRegister() {
        val arrayFiles = obtenerRegistros()
        lRegistro.clear()
        if (arrayFiles.isNotEmpty()) {
            for (element in arrayFiles) {
                lRegistro.add(ListaRegistro(element, R.drawable.ic_registros_menu))
            }
        } else {
            Toast.makeText(mContext, "Carpeta Vacia", Toast.LENGTH_LONG).show()
            lRegistro.add(ListaRegistro("Carpeta Vacia", R.drawable.ic_sin_registro))
        }
        val itemDecoration = DividerItemDecoration(mContext, DividerItemDecoration.VERTICAL)
        ContextCompat.getDrawable(mContext, R.drawable.divider)?.let {
            itemDecoration.setDrawable(it)
        }
        binding.Resgistros.apply {
            setHasFixedSize(true)
            layoutManager = GridLayoutManager(mContext, 1, GridLayoutManager.VERTICAL, false)
            adapter = RegisterListAdapter(lRegistro, menuViewModel)
            addItemDecoration(itemDecoration)
        }
    }

    fun obtenerRegistros(): List<String> {
        val item: MutableList<String> = ArrayList()
        val db = DB(mContext)
        user = getObjectBoleta()
        val folder = File(Environment.getExternalStorageDirectory().path + "/Monitoreo" + user.boleta)
        Toast.makeText(mContext, folder.path, Toast.LENGTH_LONG).show()
        if (folder.exists() && !folder.listFiles().isNullOrEmpty()) {
            folder.listFiles().also {
                if (!it.isNullOrEmpty()) {
                    for (file in (it)) if (file.path.endsWith(".csv") && db.getRecord(Archivo(file.name, user.boleta)))
                        item.add(file.name)
                }
            }
            return item
        } else {
            Toast.makeText(mContext, "No existe la carpeta del usuario", Toast.LENGTH_LONG).show()
        }
        return item
    }
}